src/BalanceTx.purs

```diff
@@ -228,117 +198,131 @@ derive instance newtypeActual :: Newtype Actual _
-      pure do
-        prebalancedTx :: Transaction <- prebalancedTx'
-        lmap ReturnAdaChangeFailure' $
-          returnAdaChange ownAddr utxoIndex' prebalancedTx
+balanceTxM :: UnbalancedTx -> QueryM (Either BalanceTxError Transaction)
+balanceTxM (UnbalancedTx { transaction: unbalancedTx, utxoIndex }) = do
+  ownAddr' <- getWalletAddress
+  collateral' <- getWalletCollateral
+  -- Saves on an extra nested case:
+  case ownAddr', collateral' of
+    Nothing, _ ->
+      pure $ Left $ GetWalletAddressError' CouldNotGetNamiWalletAddress
+    _, Nothing ->
+      pure $ Left $ GetWalletCollateralError' CouldNotGetNamiCollateral
+    Just ownAddr, Just collateral -> do
+      utxos' <- map unwrap <$> utxosAt ownAddr
+      utxoIndex' <- liftEffect $ utxoIndexToUtxo utxoIndex
+      -- Don't need to sequence, can just do as above
+      case utxos', utxoIndex' of
+        Nothing, _ ->
+          pure $ Left $ UtxosAtError' CouldNotGetUtxos
+        _, Nothing ->
+          pure $ Left $ UtxoIndexToUtxoError' CouldNotConvertUtxoIndex
+        Just utxos, Just utxoIndex'' -> do
+          let
+            -- Combines utxos at the user address and those from any scripts
+            -- involved with the contract in the unbalanced transaction. Don't
+            -- add collateral to this for now.
+            allUtxos :: Utxo
+            allUtxos = utxos `Map.union` utxoIndex''
+          -- Add hardcoded Nami 5 Ada and sign before instead of recursively
+          -- signing in loop. Could refactor this part into a separate function.
+          signedUnbalancedTx <-
+            signTransaction (addTxCollateral unbalancedTx collateral)
+              <#> note (SignTxError' $ CouldNotSignTx ownAddr)
+          case signedUnbalancedTx of
+            Left err -> pure $ Left err
+            Right sUbTx ->
+              -- After adding collateral, we need to balance the inputs and
+              -- non-Ada outputs before looping, i.e. we need to add input fees
+              -- for the Ada only collateral. No MinUtxos required:
+              -- Alternatively add this functionality back to preBalanceTxBody
+              -- and only call collateral adder initially.
+              -- Balance tx without fees:
+              case prebalanceCollateral zero allUtxos ownAddr sUbTx of
+                Left err -> pure $ Left err
+                Right sUbTx' -> do
+                  fees' <- lmap CalculateMinFeeError' <$> calculateMinFee' sUbTx'
+                  -- Balance tx with fees:
+                  let
+                    signedCollUnbalancedTx' = fees' >>= \fees ->
+                      prebalanceCollateral
+                        (fees + feeBuffer)
+                        allUtxos
+                        ownAddr
+                        sUbTx'
+                  case signedCollUnbalancedTx' of
+                    Left err -> pure $ Left err
+                    Right signedCollUnbalancedTx ->
```

Comment:
I think readability of this function can be improved by using ExceptT.

Reply:
Yes, there was a short discussion a while back about whether to use `ExceptT` - in the end, we decided against (I can link you on Slack). Another option is to just throw (JavaScript) errors but that feels unpleasant but easier to read at least.

Reply:
If we go with `ExceptT`, that might be something we want to add to the `QueryM` stack. I'll open another issue for standardizing error-handling